### PR TITLE
Context-aware API facade caller interface

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following is the next step in the road for making the API package context-aware. It adds `context.Context` available to the facade caller. I've not made every method context-aware yet, as doing that in the same PR as this is a lot of effort and probably worth doing in smaller chunks, as many tests also need to be updated.

In addition to this work I also made it possible to pass a tracer at the instantiation of the client, rather than at the level of the context. This is for two reasons:

 - For workers it makes it a lot more ergonomic to have a tracer per client.
 - It allows for a longer transition period for adding `context.Context` without re-writing all the methods.

It should be noted that the `context.Context` passed in the methods have priority over the embedded tracer, so adding them to the API can then be overridden. Leaving us with a clean transition.

----

The code is very mechanical, in fact, it was mainly done via `sed`. I've used `context.TODO()` like prior PRs so it makes it easy to locate what needs to be replaced in the future.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Tests pass and we can bootstrap.

```sh
$ juju bootstrap lxd test --build-agent
```

## Links

**Jira card:** JUJU-[XXXX]

### CODE REVIEW

**Review by hpidcock — APPROVED** [review]
Love it. Thank-you.

File: api/agent/caasapplication/client.go [hunk 1]
```diff
@@ -4,22 +4,31 @@
...
 	"github.com/juju/juju/api/base"
 	"github.com/juju/juju/rpc/params"
 )
 
+// Option is a function that can be used to configure a Client.
+type Option = base.Option
+
+// WithTracer returns an Option that configures the Client to use the
+// supplied tracer.
+var WithTracer = base.WithTracer
```
Comment (hmlanigan) [inline api/agent/caasapplication/client.go:21]: Darn it, just put together that we have 2 "base" packages inside juju. :-(

File: api/base/caller.go [hunk 2]
```diff
@@ -145,11 +145,15 @@ func NewFacadeCaller(caller APICaller, facadeName string, options ...Option) Fac
 // NewFacadeCallerForVersion wraps an APICaller for a given facade
 // name and version.
 func NewFacadeCallerForVersion(caller APICaller, facadeName string, version int, options ...Option) FacadeCaller {
+	// Derive the context from the API caller context if it's available. The
+	// default will be a noop tracer if none is found.
+	tracer, _ := coretrace.TracerFromContext(caller.Context())
```
Comment (hmlanigan) [inline api/base/caller.go:150]: Is the bool return value for TracerFromContext ever used?

File: core/trace/context.go [hunk 3]
```diff
@@ -53,6 +53,24 @@ func WithTracer(ctx context.Context, tracer Tracer) context.Context {
...
+// isn't already set on the context.
+func InjectTracerIfRequired(ctx context.Context, tracer Tracer) context.Context {
+	// If the tracer is nil, we'll just pass back the context, as that will
+	// either have a tracer or it won't. Using nil tracer could invalidate the
+	// parent one, so just send it back.
+	if tracer == nil {
+		return ctx
+	}
+
+	// If the parent tracer is parent tracer is not nil, then use that one.
```
Comment (hmlanigan) [inline core/trace/context.go:66]: `the parent tracer is parent tracer is not nil` - Typo?

**Review by hmlanigan — APPROVED** [review]
This will be interesting for terraform provider as well. Terraform passes a context around everywhere too.

Great start. Thank you for multiple PRs, even this one is long enough my browser isn't happy.

File: api/base/caller.go (hunk 2 above)
Comment (SimonRichardson) [reply api/base/caller.go:150]: Yeah, internally in the txn runner (PR https://github.com/juju/juju/pull/16402), we need to know if it's actually enabled from just the context side.

**Pull request merged**